src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelAttributes.cs

```diff
@@ -32,17 +33,56 @@ public static IEnumerable<object> GetAttributesForParameter(ParameterInfo parame
         /// <summary>
         /// Gets the attributes for the given <paramref name="property"/>.
         /// </summary>
-        /// <param name="property">A <see cref="ParameterInfo"/> for which attributes need to be resolved.
+        /// <param name="type">A <see cref="Type"/> on which the property exists.
+        /// <param name="property">A <see cref="PropertyInfo"/> for which attributes need to be resolved.
         /// </param>
         /// <returns>An <see cref="IEnumerable{object}"/> containing the attributes on the
         /// <paramref name="property"/> before the attributes on the <paramref name="property"/> type.</returns>
-        public static IEnumerable<object> GetAttributesForProperty(PropertyInfo property)
+        public static IEnumerable<object> GetAttributesForProperty(Type type, PropertyInfo property)
         {
             // Return the property attributes first.
             var propertyAttributes = property.GetCustomAttributes();
             var typeAttributes = property.PropertyType.GetTypeInfo().GetCustomAttributes();
 
-            return propertyAttributes.Concat(typeAttributes);
+            propertyAttributes = propertyAttributes.Concat(typeAttributes);
+
+            var modelMedatadataType = type.GetTypeInfo().GetCustomAttribute<ModelMetadataTypeAttribute>();
+            if (modelMedatadataType != null)
+            {
+                var modelMedatadataProperty = modelMedatadataType.MetadataType.GetRuntimeProperty(property.Name);
+                if (modelMedatadataProperty != null)
+                {
+                    var modelMedatadataAttributes = modelMedatadataProperty.GetCustomAttributes();
+                    propertyAttributes = propertyAttributes.Concat(modelMedatadataAttributes);
+
+                    var modelMetadataTypeAttributes =
+                        modelMedatadataProperty.PropertyType.GetTypeInfo().GetCustomAttributes();
+                    propertyAttributes = propertyAttributes.Concat(modelMetadataTypeAttributes);
+                }
+            }
+
+            return propertyAttributes;
+        }
+
+        /// <summary>
+        /// Gets the attributes for the given <paramref name="type"/>.
+        /// </summary>
+        /// <param name="type">A <see cref="Type"/> for which attributes need to be resolved.
+        /// <param name="attributes">A <see cref="IEnumerable{Attribute}"/> of attributes on type.
+        /// </param>
+        /// <returns>An <see cref="IEnumerable{Attribute}"/> that includes attributes on ModelMetadataType class
+        public static IEnumerable<Attribute> GetModelMetadataAttributesForType(Type type)
```

Comment:
return `IEnumerable<object>` for consistency with `GetCustomAttributes()` and other methods in this class.  does this compile as-is?